### PR TITLE
Avoid synthesizing attrs if the attr name is a non-identifier string

### LINKED ISSUE #1573: (From Fuzzer) condition this->header.find('\n') == std::string::npos has failed: Use of undeclared variable
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23typed%3Astrict%0Aattr%22%0A%22)

```ruby
#typed:strict
attr"
"
```

#### Observed output

```
Exception::raise(): ./core/Error.h:85 enforced condition this->header.find('\n') == std::string::npos has failed: Use of undeclared variable `@
` has a newline in it

Backtrace:
sorbet::core::Error::Error(sorbet::core::Loc, sorbet::core::ErrorClass, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<sorbet::core::ErrorSection, std::__1::allocator<sorbet::core::ErrorSection> >, std::__1::vector<sorbet::core::AutocorrectSuggestion, std::__1::allocator<sorbet::core::AutocorrectSuggestion> >, bool) (in libcore.so) + 331
std::__1::__unique_if<sorbet::core::Error>::__unique_single std::__1::make_unique<sorbet::core::Error, sorbet::core::Loc&, sorbet::core::ErrorClass&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<sorbet::core::ErrorSection, std::__1::allocator<sorbet::core::ErrorSection> >, std::__1::vector<sorbet::core::AutocorrectSuggestion, std::__1::allocator<sorbet::core::AutocorrectSuggestion> >, bool&>(sorbet::core::Loc&&&, sorbet::core::ErrorClass&&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, std::__1::vector<sorbet::core::ErrorSection, std::__1::allocator<sorbet::core::ErrorSection> >&&, std::__1::vector<sorbet::core::AutocorrectSuggestion, std::__1::allocator<sorbet::core::AutocorrectSuggestion> >&&, bool&&&) (in libcore.so) + 186
sorbet::core::ErrorBuilder::~ErrorBuilder() (in libcore.so) + 94
sorbet::cfg::unresolvedIdent2Local(sorbet::cfg::CFGContext, sorbet::ast::UnresolvedIdent*) (in libbuilder.so) + 1035
sorbet::cfg::CFGBuilder::walk(sorbet::cfg::CFGContext, sorbet::ast::Expression*, sorbet::cfg::BasicBlock*) (in libbuilder.so) + 2927
sorbet::cfg::CFGBuilder::buildFor(sorbet::core::Context, sorbet::ast::MethodDef&) (in libbuilder.so) + 1725
sorbet::realmain::pipeline::CFGCollectorAndTyper::preTransformMethodDef(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >) (in libpipeline.so) + 190
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapMethodDef(std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >, sorbet::core::Context) (in libpipeline.so) + 78
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) + 663
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapClassDef(std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::core::Context) (in libpipeline.so) + 151
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) + 884
std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> > sorbet::ast::TreeMap::apply<sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper>(sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper&, std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >) (in libpipeline.so) + 70
sorbet::realmain::pipeline::typecheckOne(sorbet::core::Context, sorbet::ast::ParsedFile, sorbet::realmain::options::Options const&) (in libpipeline.so) + 1599
std::__1::__function::__func<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3, std::__1::allocator<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3>, void ()>::operator()() (in libpipeline.so) + 284
sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&) (in libpipeline.so) + 786
sorbet::realmain::realmain(int, char**) (in librealmain.so) + 8499
main (in sorbet) + 11
start (in libdyld.dylib) + 1
0x00000002 (in sorbet)

Backtrace:
sorbet::cfg::unresolvedIdent2Local(sorbet::cfg::CFGContext, sorbet::ast::UnresolvedIdent*) (in libbuilder.so) + 1035
sorbet::cfg::CFGBuilder::walk(sorbet::cfg::CFGContext, sorbet::ast::Expression*, sorbet::cfg::BasicBlock*) (in libbuilder.so) + 2927
sorbet::cfg::CFGBuilder::buildFor(sorbet::core::Context, sorbet::ast::MethodDef&) (in libbuilder.so) + 1725
sorbet::realmain::pipeline::CFGCollectorAndTyper::preTransformMethodDef(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >) (in libpipeline.so) + 190
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapMethodDef(std::__1::unique_ptr<sorbet::ast::MethodDef, std::__1::default_delete<sorbet::ast::MethodDef> >, sorbet::core::Context) (in libpipeline.so) + 78
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) + 663
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapClassDef(std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::core::Context) (in libpipeline.so) + 151
sorbet::ast::TreeMapper<sorbet::realmain::pipeline::CFGCollectorAndTyper, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in libpipeline.so) + 884
std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> > sorbet::ast::TreeMap::apply<sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper>(sorbet::core::Context, sorbet::realmain::pipeline::CFGCollectorAndTyper&, std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >) (in libpipeline.so) + 70
sorbet::realmain::pipeline::typecheckOne(sorbet::core::Context, sorbet::ast::ParsedFile, sorbet::realmain::options::Options const&) (in libpipeline.so) + 1599
std::__1::__function::__func<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3, std::__1::allocator<sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&)::$_3>, void ()>::operator()() (in libpipeline.so) + 284
sorbet::realmain::pipeline::typecheck(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&) (in libpipeline.so) + 786
sorbet::realmain::realmain(int, char**) (in librealmain.so) + 8499
main (in sorbet) + 11
start (in libdyld.dylib) + 1
0x00000002 (in sorbet)

libc++abi.dylib: terminate_handler unexpectedly returned
```

#### Expected behavior

Exception does not occur.

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
Fixes #1573; `attr`, `attr_reader`, `attr_writer`, and `attr_accessor` all can take a string which tells it the name of the method to generate, but we were accepting any strings even when Ruby would reject them for not being valid. The case of the crash in #1573, we would tell it to generate an accessor named `"\n"`, and then we'd fail an ENFORCE when trying to print that string, but we could just reject any string that Ruby would reject. (Ruby is rather conservative about this: even some perfectly valid Ruby identifiers, like those that end in `?` or `!`, will be rejected by the `attr` functions.)

### Test plan

See included automated tests. Included a variation on the failing test—the original included a newline, but that causes us to generate a multi-line span that `test_corpus` cannot yet handle, so I modified the test to include a different but also invalid accessor name.

### CODE REVIEW

**Comment by aisamanra** [comment]
I was confident that this was correct, but I also double-checked this by running against pay-server and it's fine.

File: dsl/attr_reader.cc [hunk 1]
```diff
@@ -21,7 +21,19 @@ pair<core::NameRef, core::Loc> getName(core::MutableContext ctx, ast::Expression
...
+            core::NameRef nameRef = lit->asString(ctx);
+            bool validAttr =
+                absl::c_all_of(nameRef.data(ctx)->shortName(ctx), [&](char c) { return isalpha(c) || c == '_'; });
+            if (validAttr) {
+                res = nameRef;
+            } else {
+                if (auto e = ctx.state.beginError(name->loc, core::errors::DSL::BadAttrArg)) {
+                    // we can't necessarily repeat the name here because it might have newlines in it, in which case our
+                    // expected invariant that all error lines are a single line will break
+                    e.setHeader("Invalid string provided for attribute name");
```
Comment (DarkDimius) [inline dsl/attr_reader.cc]: `Ruby` uses 
```
invalid attribute name `{name}`
```
, let's do the same?

File: dsl/attr_reader.cc [hunk 2]
```diff
@@ -21,7 +21,19 @@ pair<core::NameRef, core::Loc> getName(core::MutableContext ctx, ast::Expression
             ENFORCE(loc.source(ctx).size() > 1 && loc.source(ctx)[0] == ':');
             loc = core::Loc(loc.file(), loc.beginPos() + 1, loc.endPos());
         } else if (lit->isString(ctx)) {
-            res = lit->asString(ctx);
+            core::NameRef nameRef = lit->asString(ctx);
+            bool validAttr =
+                absl::c_all_of(nameRef.data(ctx)->shortName(ctx), [&](char c) { return isalpha(c) || c == '_'; });
```
Comment (elliottt) [inline dsl/attr_reader.cc]: Won't this reject strings like `"foo7"`?

**Comment by aisamanra** [comment]
I switched to using a regex to match the name, because that's probably clearer than using an explicit loop. Happy to hear FUD about using regexes in this way.

File: dsl/attr_reader.cc (hunk 2 above)
Comment (elliottt) [reply dsl/attr_reader.cc]: The regex fixes this, thanks!

File: dsl/attr_reader.cc [hunk 3]
```diff
@@ -21,7 +23,20 @@ pair<core::NameRef, core::Loc> getName(core::MutableContext ctx, ast::Expression
             ENFORCE(loc.source(ctx).size() > 1 && loc.source(ctx)[0] == ':');
             loc = core::Loc(loc.file(), loc.beginPos() + 1, loc.endPos());
         } else if (lit->isString(ctx)) {
-            res = lit->asString(ctx);
+            core::NameRef nameRef = lit->asString(ctx);
+            auto shortName = nameRef.data(ctx)->shortName(ctx);
+            bool validAttr =
+                regex_match(shortName.begin(), shortName.end(), basic_regex("([:alpha:]|_)([:alnum:]|_)*"));
```
Comment (DarkDimius) [inline dsl/attr_reader.cc]: c++ regexes are notoriously bad. Please use an explicit loop :-)

Comment (DarkDimius) [reply dsl/attr_reader.cc]: See https://github.com/sorbet/sorbet/commit/81b567ddd90c087d77d948c10aef89d40f5343f8 for prior example

**Comment by aisamanra** [comment]
Switched away from the regex.

**Pull request merged**